doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,919 @@
+Found 1 error in 1 file (checked 1 source file)
+```
+
+Because Penguin does not implement the `SupportsFly` Protocol is called out as an incompatible type. However, any type
+that is consistent with, or implements all the methods defined by the `SupportsFly` protocol, is acceptable to pass
+into `ascend()`.
+
+A key difference between Protocols and abstract classes (abc) is that protocols work at typing time while abcs work at
+runtime. A class does not need to _inherit_ from Protocol like an abc, the type checker will simply enforce its usage at
+typing time.
+
+A common Python SDK Protocol type we use
+is [azure.core.credentials.TokenCredential](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/azure/core/credentials.py)
+. TokenCredential requires that a type implements the `get_token` method for it to be consistent with the Protocol.
+
+See more on Protocols in the [reference documentation](https://docs.python.org/3/library/typing.html#typing.Protocol).
+
+#### Use runtime_checkable to do simple, runtime structural checks
+
+If a Protocol is marked with @runtime_checkable, it can be used with `isinstance()` and `issubclass()` at runtime.
+
+```python
+# from typing import runtime_checkable  Python >=3.8
+from typing_extensions import runtime_checkable
+
+
+@runtime_checkable
+class SupportsFly(Protocol):
+    def fly(self) -> None: ...
+
+
+assert isinstance(Plane, SupportsFly)  # True
+assert isinstance(Penguin, SupportsFly)  # False
+```
+
+Because this is a runtime check, only the presence of the required methods defined by the Protocol is checked -- not
+their type signatures.
+
+> Note that runtime_checkable is backported to older versions of Python by using typing_extensions
+
+### Use typing.overload to overload a function
+
+`typing.overload` allows for annotating different combinations of arguments for a function. This is useful when the
+return type might depend on the type or types of parameters.
+
+Consider a function which takes in different input types which then inform the output type:
+
+```python
+from typing import Union
+
+
+def analyze_text(text: str, analysis_kind: Union[SentimentAnalysis, EntityRecognition, LanguageDetection]) -> Union[
+    SentimentResult, EntityRecognitionResult, LanguageDetectionResult]:
+    return _analyze(text, analysis_kind)
+```
+
+We can annotate this function with overloads to help inform the type checker (and intellisense) which input maps to
+which output type.
+
+```python
+from typing import overload
+
+
+@overload
+def analyze_text(text: str, analysis_kind: LanguageDetection) -> LanguageDetectionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: EntityRecognition) -> EntityRecognitionResult:
+    ...
+
+
+@overload
+def analyze_text(text: str, analysis_kind: SentimentAnalysis) -> SentimentResult:
+    ...
+
+
+def analyze_text(text, analysis_kind):  # the actual implementation doesn't need to be annotated
```

Comment:
Here you say "doesn't **need to** be". Wondering if this should be re-worded as **"shouldn't be"**? When I was testing this, I saw that actually annotating the implementation results in errors that would otherwise not be there.

For example, this does not result in mypy errors.
```python
@overload
def get_zero(a: Dict) -> str:
    ...
@overload
def get_zero(a: List) -> str:
    ...
def get_zero(a):
    try:
        zero = a[0]
    except AttributeError:
        zero = a.pop("0", "val")
    return zero
```

But annotating the implementation:
```python
def get_zero(a: Union[Dict, List]) -> str:
    try:
        zero = a[0]
    except AttributeError:
        zero = a.pop("0", "val")
    return zero
```
Results in:
```
encode_and_decode_event_data_message.py:102: error: Too many arguments for "pop" of "list"
encode_and_decode_event_data_message.py:102: error: Argument 1 to "pop" of "list" has incompatible type "str"; expected "SupportsIndex
```